### PR TITLE
EC-773

### PR DESCRIPTION
**Pull request by SunFrost**

### CODE REVIEW

File: src/main/scala/org/encryfoundation/prismlang/compiler/StaticAnalyser.scala [hunk 1]
```diff
@@ -190,51 +208,55 @@ case class StaticAnalyser(initialScope: ScopedSymbolTable, types: TypeSystem) ex
...
+        throw SemanticAnalysisException(s"String max length overflow (${value.length} > ${Constants.StringMaxLength})")
       string
 
     /** Make sure given `value` does not overflow `Long.MaxSize`. */
     case int @ Expr.IntConst(value) =>
-      if (value > Long.MaxValue) error(s"int64 max size overflow ($value > ${Long.MaxValue})")
+      if (value > Long.MaxValue) throw SemanticAnalysisException(s"int64 max size overflow ($value > ${Long.MaxValue})")
       int
+
+    /** Make sure given `value` does not overflow `Long.MaxSize`. */
```
Comment (GRYE) [inline src/main/scala/org/encryfoundation/prismlang/compiler/StaticAnalyser.scala]: I think, it should be `Byte.MaxValue`

File: src/main/scala/org/encryfoundation/prismlang/lib/predefined/base/BaseFunctionHolder.scala [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+  val cost: Int = 10
+
+  def asFunc: PFunctionPredef = PFunctionPredef(args, body)
+
+  val args: IndexedSeq[(String, Types.PType)] = IndexedSeq("input" -> Types.PCollection.ofByte)
+
+  val baseFunc: Array[Byte] => String
+
+  val body: Seq[(String, PValue)] => Either[PFunctionPredef.PredefFunctionExecFailure.type, Any] = (pArgs: Seq[(String, PValue)]) => {
```
Comment (GRYE) [inline src/main/scala/org/encryfoundation/prismlang/lib/predefined/base/BaseFunctionHolder.scala]: Why `Either[PFunctionPredef.PredefFunctionExecFailure.type, Any]`, and not `Either[PFunctionPredef.PredefFunctionExecFailure.type, ]`

File: src/main/scala/org/encryfoundation/prismlang/lib/predefined/hash/HashFunctionHolder.scala [hunk 3]
```diff
@@ -4,22 +4,22 @@ import org.encryfoundation.prismlang.core.Types
...
 
-  protected def bodyValue(hash: Array[Byte] => Array[Byte]): Seq[(String, PValue)] => Either[PFunctionPredef.PredefFunctionExecFailure.type, Any] = (varargs: Seq[(String, PValue)]) => {
-    val validNumberOfArgs = varargs.size == args.size
-    val validArgTypes = varargs.zip(args).forall { case ((_, value), (_, tpe)) => value.tpe == tpe }
-    if (validNumberOfArgs && validArgTypes) {
-      val fnArgs = varargs.map(_._2.value.asInstanceOf[Array[Byte]])
-      Right(hash(fnArgs.head).toList)
+  val hashFunc: Array[Byte] => Digest
+
+  val body: Seq[(String, PValue)] => Either[PFunctionPredef.PredefFunctionExecFailure.type, Any] = (pArgs: Seq[(String, PValue)]) => {
```
Comment (GRYE) [inline src/main/scala/org/encryfoundation/prismlang/lib/predefined/hash/HashFunctionHolder.scala]: Again, why `Any`?

File: src/test/scala/org/encryfoundation/prismlang/evaluator/OperatorsTypeCompatibilitySpec.scala [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+  property("binary operators shouldn't compile with different types") {
+    checkBinaryOperators(List(Operator.Add, Operator.Sub, Operator.Mult, Operator.Div, Operator.Mod, Operator.Pow),
+      values1, values2, binExpr)
+  }
+
+  property("compare operators shouldn't compile with different types") {
+    checkBinaryOperators(List(CompOp.GtE, CompOp.Gt, CompOp.Lt, CompOp.LtE), values1, values2, compareExpr)
+  }
+
+//  property("Eq NotEq shouldn't compile with different types") {
```
Comment (GRYE) [inline src/test/scala/org/encryfoundation/prismlang/evaluator/OperatorsTypeCompatibilitySpec.scala]: Why are these tests commented?

File: src/main/scala/org/encryfoundation/prismlang/lib/predefined/base/BaseFunctionHolder.scala [hunk 5]
```diff
@@ -0,0 +1,26 @@
+package org.encryfoundation.prismlang.lib.predefined.base
+
+import org.encryfoundation.prismlang.core.Types
+import org.encryfoundation.prismlang.core.wrapped.PFunctionPredef.PredefFunctionExecFailure
+import org.encryfoundation.prismlang.core.wrapped.{PFunctionPredef, PValue}
+import org.encryfoundation.prismlang.lib.predefined.BuiltInFunctionHolder
+
+trait BaseFunctionHolder extends BuiltInFunctionHolder {
+
+  val cost: Int = 10
```
[the remark this answers is not in the document]
Comment (SunFrost) [reply src/main/scala/org/encryfoundation/prismlang/lib/predefined/base/BaseFunctionHolder.scala]: saw in 

object Base58decode extends BuiltInFunctionHolder {

  val name: String = "decode"

  val cost: Int = 10
----------------------------------
How to calculate it ?

File: src/test/scala/org/encryfoundation/prismlang/evaluator/OperatorsTypeCompatibilitySpec.scala (hunk 4 above)
Comment (SunFrost) [reply src/test/scala/org/encryfoundation/prismlang/evaluator/OperatorsTypeCompatibilitySpec.scala]: Tests found some bugs, but these fixes have not yet been made.
Some tests was сommented that would pass in trevis

Comment (SunFrost) [reply src/test/scala/org/encryfoundation/prismlang/evaluator/OperatorsTypeCompatibilitySpec.scala]: I added these bugs to tasks

**Pull request merged**